Add blog: Optimizing for Almost Sorted Data: Sort Pushdown in Apache DataFusion (#186)
apache/datafusion-site: content/blog/2026-07-05-sort-pushdown.md

---
layout: post

> title: Sort Pushdown in DataFusion: Skip Sorts, Skip Decode, Skip I/O

Reviewer: What do you think about changing the title of the post to something like
```suggestion
title: Optimizing for Almost Sorted Data: Sort Pushdown in Apache DataFusion
```

( I am absolutely trying to increase readership by making this sound more appealing than to the datafusion die hards)

Author: Good suggestion! Committed this suggestion.

---
layout: post
title: Sort Pushdown in DataFusion: Skip Sorts, Skip Decode, Skip I/O
date: 2026-07-05
author: Qi Zhu and Andrew Lamb
categories: [performance]
---

<!--
{% comment %}
Licensed to the Apache Software Foundation (ASF) under one or more
contributor license agreements.  See the NOTICE file distributed with
this work for additional information regarding copyright ownership.
The ASF licenses this file to you under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

[TOC]

> *[Qi Zhu](https://github.com/zhuqi-lucas), [Massive](https://www.massive.com/); [Andrew Lamb](https://github.com/alamb), [InfluxData](https://www.influxdata.com/)*

Author: Thanks @alamb I added your info, is it right? Let me know if it's not right.

Reviewer: Yes, thank you

Reviewer: Yes, thank you

Reviewer: Yes, thank you

---
layout: post
title: Sort Pushdown in DataFusion: Skip Sorts, Skip Decode, Skip I/O
date: 2026-07-05
author: Qi Zhu
categories: [performance]
---

<!--
{% comment %}
Licensed to the Apache Software Foundation (ASF) under one or more
contributor license agreements.  See the NOTICE file distributed with
this work for additional information regarding copyright ownership.
The ASF licenses this file to you under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

[TOC]

*Qi Zhu, [Massive](https://www.massive.com/)*

> **[Apache DataFusion] now automatically takes advantage of sortedness in the

Reviewer: I really like the intro now

One nit is you can make maybe even stronger by removing the `now` so this post isn't quite as tied to some historic plan / process, but is describing a universal set of optimziations

```suggestion
**[Apache DataFusion]  automatically takes advantage of sortedness in the
```

Author: Done — dropped `now`.

data — even when the data is only *partially* sorted, and even when
DataFusion has not been told about the ordering ahead of time.** This post
explains why that matters and walks through how DataFusion achieves it,
through a combination of plan-time sort pushdown, runtime scan reordering,
and mid-scan row-group pruning driven by [dynamic filters][dyn-filters-blog].

[Apache DataFusion]: https://datafusion.apache.org/
[dyn-filters-blog]: https://datafusion.apache.org/blog/2025/09/10/dynamic-filters/

## Why sort pushdown matters

> Many real datasets are at least partly sorted on disk:

Reviewer: Maybe use terminology less connected to past tech:

```suggestion
Many real datasets are at least partly sorted when stored:
```

Author: Done — `sorted on disk` → `sorted when stored`.

- Time-series files are written in ingestion-time order.
- Event logs are sharded and sorted by event id.
- Partitioned tables have a natural ordering by partition key.
- Modern data lakes based on [Apache Iceberg] and similar formats
  often have to work with data **as it was written** — resorting the

>   whole table isn't an option.

Reviewer: This item seems a little different than the others that are explaining usescase when data is sorted on disk

maybe something more like

```suggestion
- Modern data lakes based on [Apache Iceberg] and similar formats
  often store data in the order  **it was written**, and resorting the
  data is prohibitively expensive for many workloads.
```

Author: Done — adopted suggested Iceberg bullet.

But that "pre-existing sortedness" is only useful if the query engine can
**notice** it and **use** it. Two common failure modes:

1. The engine doesn't know about the ordering — the writer didn't set

>    Parquet `sorting_columns`, and the table definition doesn't include a

Reviewer: recommend a link to sorting columns defintion: https://github.com/apache/parquet-format/blob/8a5e04bdecf100e8e981daacfa117e8b5aadacb9/src/main/thrift/parquet.thrift#L1044

Also maybe make it clear that the with order thing is a datafusion thing

instead of 

"and the table definition doesn't include a..."

Something like

"or the system wasn't informed about the ordering with a mechanism such as DataFusions `WITH_ORDER` clause"

Author: Done — linked Parquet `sorting_columns` to thrift; clarified `WITH ORDER` as DataFusion-specific.

   [`WITH ORDER`](https://datafusion.apache.org/user-guide/sql/ddl.html#create-external-table) clause.

> 2. The engine knows the *per-file* ordering, but the file *listing* on

Reviewer: I don't have a great suggestion here, but this per file ordering vs listing ordering seems very specific to DataFUsion -- if we can find some way to make that clear it might be easier to understand

Author: Done — noted the per-file vs listing distinction is a DataFusion ListingTable concept.

   disk is in a different order, so global sortedness can't be proven at
   plan time.

In both cases, an `ORDER BY` or `ORDER BY ... LIMIT N` query pays the
cost of a full external `SortExec` — a pipeline-blocking operator that
must see every input row before emitting anything, dominating both

> latency and peak memory on large scans.

Reviewer: I recommend making this more generic as it applies to all systems, not just DataFusion

```suggestion
In both cases, an `ORDER BY` or `ORDER BY ... LIMIT N` query pays the
cost of a full sort — a pipeline-blocking operator that
must buffer every input row before emitting anything, dominating both
latency and peak memory on large scans.
```

Author: Done — `full external SortExec` → `full sort`; `see` → `buffer`.

Min/max statistics used for *predicate* pushdown are well-known and
widely implemented across databases. Using them to *reason about sort

> order* — deleting redundant sorts, biasing scan order toward the

Reviewer: ```suggestion
order* — deleting redundant sorts and biasing scan order toward the
```

Author: Done — `sorts, biasing` → `sorts and biasing`.

most-promising data — is less common. This post is about how DataFusion
does the latter.

[Apache Iceberg]: https://iceberg.apache.org/

## What DataFusion could already do — and what was missing

DataFusion has always been able to skip the sort in the **exact** case,
using the machinery covered in [@akurmustafa's earlier post on
ordering analysis][ordering-analysis]: when the table definition
declares an ordering (via `WITH ORDER` or Parquet `sorting_columns`)
**and** the on-disk file listing already matches that order, the
existing `EnsureRequirements` rule sees that the scan's
`output_ordering` satisfies the request and **removes the redundant
`SortExec`** entirely.

This post is about **everything else** — the messier real-world cases
where sortedness exists but isn't provable up front:

- Files listed in the "wrong" order on disk (each file internally
  sorted, but the listing doesn't match).
- Declared ordering with **overlapping** ranges across files.
- **No** declared ordering at all.
- `ORDER BY ... DESC` on ASC-sorted data.

Three complementary techniques close each gap:

> 1. **Statistics-based sort elimination** (`Exact` path). Extend the

Reviewer: as above, i think using "sort" rather than `SortExec` would make it easier to see that this is not something only applicable to datafusion

Author: Done — `SortExec` → `sort` in the three-technique bullets.

   optimizer to prove ordering from min/max statistics after
   reordering the file list, then delete the `SortExec` entirely.
2. **Runtime scan reorder** (`Inexact` path). Keep the `SortExec`, but

>    bias scan order so the *most-promising* data is read first —

Reviewer: ```suggestion
   bias  the order files are scanned so the *most-promising* data is read first —
```

Author: Done — applied suggested phrasing.

   `TopK`'s [dynamic filter][dyn-filters-blog] tightens quickly and
   downstream data is pruned by statistics before it's read.

> 3. **Runtime row-group dynamic pruning** ([#22450]). Inside the

Reviewer: This refers to "TopK threshold" before that concept is introduced -- if we could describe this in more general terms here it might be better -- maybe something like "parquet decoder loop, re-check 
dynamic predicate pruning if they have gotten more precise due to additional runtime information."

or something like this

Author: Done — rewrote in generic terms, linked to the dynamic-filters blog.

   parquet decoder loop, re-check the live `TopK` threshold at every
   row-group boundary and physically remove pruned row groups before
   any bytes are fetched.

Together these compose into a **three-layer pruning stack**
(file-level, row-group-level, row-level), all driven by the same

> `TopK` dynamic filter. Headline results:

Reviewer: again here it is somewhat confusing to refer to TopK dynamic filtering without defining it -- maybe if you just included a link to the dynamic pruning blog post on the first reference it would be easier

Author: Done — rewrote in generic terms.

  `LIMIT N` becomes a static fetch on the source (the reader stops the
  moment N rows are emitted).
- **`Inexact`** — the optimizer believes the output is probably ordered
  but cannot prove it. Downstream operators like
  [`SortPreservingMergeExec`](https://docs.rs/datafusion-physical-plan/latest/datafusion_physical_plan/sorts/sort_preserving_merge/struct.SortPreservingMergeExec.html) can still benefit, but the
  explicit `SortExec` stays for correctness. In this case `TopK`'s
  [dynamic filter][dyn-filters-blog] tightens as the heap fills, and
  data whose min/max cannot beat the threshold is pruned before it is
  fully read.

For example, given a query that returns the 10 most recent trades:

```sql
SELECT ts, symbol, amount FROM trades ORDER BY ts DESC LIMIT 10;
```

- With no ordering knowledge, DataFusion scans everything and uses a
  `TopK` heap to keep the running best 10.
- With **`Exact`** ordering, DataFusion drops the sort entirely and
  stops reading after emitting 10 rows.
- With **`Inexact`** ordering, the `SortExec` stays but scans start

>   from the most-promising data, so the `TopK` threshold tightens fast

Reviewer: ```suggestion
  from the most-promising data, so the `TopK` threshold is more likely to tighten fast
```

Author: Done — applied suggested phrasing.

  and the rest is pruned by statistics.

The optimizer rule that upgrades a scan from `Unsupported` to
`Exact`/`Inexact` — and that removes the resulting redundant
`SortExec` — is [`PushdownSort`](https://github.com/apache/datafusion/blob/main/datafusion/physical-optimizer/src/pushdown_sort.rs). `PushdownSort`

> runs late, after `EnsureRequirements` has finalised the plan shape.

Reviewer: the fact that pushdown sort runs late seems like it is an irreelevant detail for this expanation -- we could probably remove it and shorten the cotnent.

Likewise the details that about how it walks the plan tree -- I don't think that adds anything more to the narrative and could be removed

Author: Done — dropped the `runs late after EnsureRequirements` and plan-walk detail.

It walks each `SortExec`, asks the child leaf via `try_pushdown_sort`
which flavour the source can produce, and rewrites accordingly.

## The `Exact` Path · Sort Elimination via Statistics

<img src="/blog/images/sort-pushdown/phase1-file-reorder.svg" alt="File reorder: rearranging files within a partition by min/max statistics so the file list is in range order" width="100%" class="img-fluid" /><br/>
*Figure: file reorder by per-file `min/max` puts the file list in range
order without touching file contents.*

DataFusion could already recognize the *exact* sortedness case (declared
ordering + matching on-disk file list). The new capability is recognizing
sortedness when the **file list is in the wrong order** on disk, using
the min/max statistics that the Parquet writer already stored per row
group. Implemented across two PRs on `PushdownSort`:
[apache/datafusion#19064][#19064] (rule scaffolding), and

> [apache/datafusion#21182][#21182] (stats-based file reorder).

Reviewer: ```suggestion
DataFusion could already recognize the *exact* sortedness case (declared
ordering + matching on-disk file list). It can now also recognize
sortedness when the **file list is in the wrong order** on disk, by using
the min/max row group statistics stored in the Parquet file (relevant PRs:
[apache/datafusion#19064][#19064] (rule scaffolding), and
[apache/datafusion#21182][#21182] (stats-based file reorder))
```

Author: Done — adopted suggestion verbatim.

For example, consider three files `a.parquet`, `b.parquet`,
`c.parquet`. Each is internally sorted by `ts` and declares
`WITH ORDER (ts ASC)`, but they were written by different jobs and end
up listed alphabetically on disk (which does *not* match sort order).
The old machinery has no way to prove global sortedness, so an
`ORDER BY ts` query pays for a full external sort even though the
underlying data is already sorted.

`PushdownSort` fixes this in three steps at the file-scan node:

1. **Sort the file list by per-file `min`** on the sort column.
2. **Check adjacency**: does `file[i].max ≤ file[i+1].min` hold for
   every adjacent pair? If yes, the sorted file list produces a globally
   sorted stream.
3. **Upgrade the source's ordering claim to `Exact`** and remove the
   surrounding `SortExec`.

<img src="/blog/images/sort-pushdown/phase2-stats-overlap.svg" alt="Detecting non-overlapping ranges via min/max statistics" width="100%" class="img-fluid" /><br/>
*Figure: after reorder, the left case has non-overlapping ranges (safe
to upgrade to `Exact`); the right case has overlaps (upgrade skipped,
falls through to the `Inexact` path).*

> Two conservative bail-outs: (a) sort keys must be plain columns

Reviewer: FWIW this can probably also be removed (or moved to a footnote) as an implementation detail and to shorten the article -- the section explains the optimization well without these caveats

Author: Done — removed the two conservative bail-outs paragraph.

(`ORDER BY date_trunc('hour', ts)` doesn't qualify — no per-file min/max
for the function output), and (b) sort columns must be null-free, so
`NULLS FIRST`/`NULLS LAST` semantics are preserved across file
boundaries. The overlap case falls through to the `Inexact` path
covered later.

### `BufferExec` · a subtle multi-partition side effect

<img src="/blog/images/sort-pushdown/buffer-exec-stall.svg" alt="SPM stalls when SortExec is removed in multi-partition plans" width="100%" class="img-fluid" /><br/>
*Figure: removing the per-partition `SortExec` leaves the top-of-plan
merge (`SortPreservingMergeExec`) directly consuming raw I/O; a stall
on any partition stalls the whole plan.*

Removing the `SortExec` looked like a pure win, but the first
multi-partition benchmarks showed something counter-intuitive: **some
queries got slower**. The root cause is that the removed `SortExec`
was doing two jobs — sorting *and* implicitly buffering. Each

> per-partition `SortExec` runs as its own task, greedily draining its

Reviewer: in my opinion, this will be confusing to many poeple as they won't know what a "per-partition" SortExec is -- it is a DataFusion specific thing

I suggest either
1. explain more how the SortExec works for multiple partitions or
2. (preferred) make this more generic and remove the details (remove straring at "Each per-partition...")

Instead of explaining the implementation details I suggest make the description higher level and refer readers to the PR for details. 

Maybe something like

> Removing the `SortExec` looked like a pure win, but the first
multi-partition benchmarks showed something counter-intuitive: **some
queries got slower**. The root cause is that the removed `SortExec`
was doing two jobs — sorting *and* implicitly buffering.
The fix was to introduce explicit buffering in certain plans, see PR XYZ for details) as shown in 
the following figure
...
(Figure)

Author: Done — high-level rewrite; details moved to PR #21426 link.

any single partition.

Once the `SortExec` is deleted, the merge sits directly on the raw
parquet streams. It's a lazy consumer — a k-way merge must see the
head row from every input before deciding which to emit. A stall in
*any one* partition now stalls the entire merge.

<img src="/blog/images/sort-pushdown/buffer-exec.svg" alt="BufferExec replaces the deleted SortExec with a bounded streaming buffer per partition" width="100%" class="img-fluid" /><br/>
*Figure: `BufferExec` is inserted where the `SortExec` used to live —
same greedy per-partition prefill, but no blocking sort.*

The fix is [`BufferExec`](https://github.com/apache/datafusion/blob/main/datafusion/physical-plan/src/buffer.rs): a bounded per-partition
prefill buffer that plays the same "greedy parallel I/O driver" role
the `SortExec` implicitly did. No sort, no blocking, and strictly
less memory than the `SortExec` it replaces. The capacity is bounded
(default 1 GB, configurable via
[`sort_pushdown_buffer_capacity`](https://github.com/apache/datafusion/pull/21426)) and grows via the
global memory pool, so it back-pressures the source instead of
OOMing.

> ### Benchmark: `sort_pushdown` suite

Reviewer: I think it would help to add an explanatory description here about what this benchmark suite is all about -- something like 

"To evaluate the effectiveness of eliminating sorts using statistics, we ran DataFusion's `sort_pushdown` benchmark suite"

Author: Done — added intro sentence explaining the `sort_pushdown` suite.

`--partitions 1`, versus `main`:

| Query                                       | Before  | After   | Speedup  |
| ------------------------------------------- | -------:| -------:| -------: |
| Q1 — `ORDER BY key` (full scan)             | 259 ms  | 122 ms  | **2.1×** |
| Q2 — `ORDER BY key LIMIT 100`               |  80 ms  |   3 ms  | **27×**  |
| Q3 — `SELECT * ORDER BY key`                | 700 ms  | 313 ms  | **2.2×** |
| Q4 — `SELECT * ORDER BY key LIMIT 100`      | 342 ms  |   7 ms  | **49×**  |

- **Full-scan queries (Q1, Q3)** save the cost of the sort itself
  (~½ end-to-end latency for in-memory sorts).
- **`LIMIT` queries (Q2, Q4)** benefit dramatically because deleting
  the `SortExec` turns `LIMIT N` into a **static fetch** on the source —
  the reader stops after N rows. A 342 ms full-file scan collapses
  into a 7 ms K-row read.

## The `Inexact` Path · Runtime Reorder for `TopK` and `DESC`

Stats-based sort elimination handles the `Exact` upgrade — strong
correctness, sort elimination — but only when the table has a
declared `output_ordering` *and* the files are provably
non-overlapping after sorting by min. Three classes of queries

> fall outside that window:

Reviewer: I think you can make this more concise and thus understandable by removing some of the detail and the bullets

Something like

> Stats-based sort elimination handles the case when the table has a
> declared `output_ordering` *and* the files are provably
> non-overlapping after sorting by min. However when these cases
> do not apply, a full sort is still overkill. The parquet metadata ...

Author: Done — condensed to a short paragraph, dropped the three-bullet enumeration.

* **Unsorted data** — no `WITH ORDER`, no parquet `sorting_columns`.
  The `Exact` upgrade cannot fire because there is no ordering
  claim to upgrade.
* **Overlapping ranges** — files written by different ingestion
  jobs share time windows. The `Exact` upgrade keeps the `SortExec`
  because the global ordering can't be proven, even though the
  files often do contain large stretches of in-order data.
* **`ORDER BY ... DESC` on ASC-sorted data** — flipping iteration
  at the row-group level emits "RGs descending × rows ascending",
  close to the requested order but not strictly DESC, so the
  `SortExec` has to stay for correctness.

For all three, a full external `SortExec` is overkill. The parquet
metadata is right there, and reading the *most-promising* data
first lets `TopK`'s dynamic filter threshold tighten quickly so the
rest gets pruned. Runtime reorder wires that up by generalising
the `Inexact` path the rule introduced.

> ### When Inexact fires

Reviewer: I recomment removing this heading as it is covered by the text below

Author: Done — removed the `### When Inexact fires` heading.

<img src="/blog/images/sort-pushdown/pr21956-decision.svg" alt="try_pushdown_sort decision tree: Exact, Inexact, or Unsupported" width="100%" class="img-fluid" /><br/>

> *Figure: for each `SortExec`, the leaf source returns `Exact` (drop

Reviewer: I found this wording slight confusing -- the "leaf source" of what? Maybe it would be clearer if it were something like

"for each `SortExec` that `SortPushdown` tries to push down into the scan, the data source returns `Exact`, ...."

Author: Done — clarified the caption per suggestion.

the sort), `Inexact` (bias the scan and keep the sort), or
`Unsupported`.*

The Inexact verdict fires when either of two independent signals is
true:

- **Stats-based reorder available**: the leading sort key is a plain
  column in the file schema, so the scan can sort files and row
  groups by `min(col)` from Parquet statistics.
- **Reverse satisfies the request**: the source's declared ordering,
  when reversed, satisfies what the query asks for. This uses
  DataFusion's [equivalence-properties][ordering-analysis] reasoning
  and covers function monotonicity (`ts DESC` declared, `date_trunc('day', ts) ASC`
  requested), constants inferred from filters, and multi-column
  composite orderings.

> ### How the scan reorders data

Reviewer: This section might be easier to understand with a bit more of a transition:

> If SortPushdown determines  `Inexact` applies to the data source, and the
> source supports `Inexact` further optimizations are possible. For example, the 
> parquet opener applies ....

Author: Done — added the transition sentence.

<img src="/blog/images/sort-pushdown/pr21956-runtime-pipeline.svg" alt="Runtime reorder pipeline: file reorder, RG reorder, then optional reverse" width="100%" class="img-fluid" /><br/>
*Figure: the parquet opener applies file-level reorder → row-group-level
reorder → optional iteration reverse.*

The parquet opener applies up to three composable steps at query start:

> 1. **File-level reorder** — across a shared work-stealing queue, the

Reviewer: the "shared workstealing queue" doesn't seem relevant to how files are reordered and it just make it harder to understand this sentence -- I recommend removing it

Author: Done — dropped `across a shared work-stealing queue`.

   file list is sorted by `min(col)`, so the most-promising file is
   picked first across all partitions.
2. **Row-group-level reorder** — once a file is opened, its row groups
   are sorted by `min(col)`.
3. **Iteration reverse** — flip row-group iteration order for `DESC`

>    requests (and for the reverse-satisfies-the-request cases above).

Reviewer: I think you can remove "(and for the reverse-satisfies-the-request cases above)." and not lose any of the main idea

Author: Done — dropped the reverse-satisfies parenthetical.

> ### File-level early stop already works

Reviewer: I recommend removing the "already works" here -- so this blog is mostly about the current DataFusion capabilities, rather than the order of what was added

So like

### File-level early stop

Author: Done — renamed to `File-level early stop`.

<img src="/blog/images/sort-pushdown/desc_walk_file.png" alt="Tier 1 file-level reorder with early stop via file_pruner" width="100%" class="img-fluid" /><br/>

> *Figure: after file reorder, low-value files at the tail of the queue

Reviewer: it might help here to name explcitly the `low-value` files to  help readers understand the figure -- "low-value files (`file_d` and `file_c`). Also maybe it is worth explicitly pointing out that "low-value" means "sort key values are smaller for ASC wueries"

Author: Done — named `file_d` / `file_c` and defined `low-value` (smaller sort key for ASC).

are cut by the file-level pruner before they are ever opened — no
metadata I/O.*

Once files are ordered "most-promising first", `TopK`'s heap fills
quickly and its dynamic filter threshold tightens. Low-value files at
the tail of the queue are then checked against the live threshold
by the [`FilePruner`](https://github.com/apache/datafusion/blob/main/datafusion/pruning/src/file_pruner.rs) before they are ever opened —
never loading their footer, page index, or any data.

> ### Row-group-level: the gap [#22450] fills

Reviewer: I rcommend calling this section "Row Group Filter early stop" for consistency with the section above and remove the link to #22450 (put that in the text)

THe reason I think this is more readable is that it removes a level of indirection -- as written readers now need to keep the "#22450" concept in their head and mentally translate it it to row group level pruning. Avoiding that translation step makes the content easier to understand

Author: Done — renamed to `Row Group Filter early stop`; moved PR #22450 into body.

> <img src="/blog/images/sort-pushdown/desc_walk_rg.png" alt="Tier 2 RG-level reorder — filter column still read for every RG pre-#22450" width="100%" class="img-fluid" /><br/>

Reviewer: this image says "Part 4" which is a little confusing as well as also referring to #22450 which will not make sense to the average reader I don't think

Author: Done — rewrote alt text; dropped `Tier 2` / `pre-#22450`.

*Figure: inside a file, the first row group tightens the threshold —
subsequent row groups have their projection columns short-circuited,
but the filter column still has to be read to discover that no rows
qualify.*

> Inside a file, the story is almost identical — but with one gap.

Reviewer: this only applies when filter pushdown is on, right? I think you could simplify this section by escribing this at a high level rather than details 

Something like

> by reordering row groups so the ones most likely to contain the minimum value are read first
> makes the exsiting TopK dynamic filters +  filter_pushdown optimization (link to the 
> late materialization arrow blog) more effective, and they can often rule out entire row group
> after evaluating the filter columns, and avoid even fetch projection columns.  
> This optimization was added in #22450

Author: Done — replaced with a high-level paragraph linking to late-materialization blog.

After the first row group fills the heap, subsequent row groups
whose values can't beat the threshold evaluate to an empty
`RowSelection`, and arrow-rs's reader short-circuits: no projection
columns fetched, no decompress, no decode.

> However, **the filter column still gets read for every row group**,

Reviewer: This paragraph could also be rephrased as the introduction of the next section (and again I would avoid using a PR number to name the concept)

Author: Done — merged that paragraph into the next section intro.

because the dynamic filter has to be evaluated row-by-row to
*discover* that no rows survive. On a large file with many row
groups, that's a meaningful tax — most of which is redundant, since
metadata alone could have proven the row group unwinnable. Closing
that gap is what [#22450] does.

> ## #22450 · Runtime Row-Group Dynamic Pruning

Reviewer: "Runtime Row-Group Dynamic Pruning"

Author: Done — renamed to `Runtime Row-Group Dynamic Pruning`.

> The merge that just landed — [apache/datafusion#22450][#22450] —

Reviewer: Rather than focus on #22450 I think it would be clearer to read if this sentence was framed more like "DataFusion now also re-checks the dynamic filter **at every row-group boundary** ....

Author: Done — reframed as `DataFusion now also re-checks…`.

re-checks the dynamic filter **at every row-group boundary** inside
an open file, converts the live threshold into a fresh

> `PruningPredicate`, and physically removes any row group whose

Reviewer: It might also be good to point out this only happens if the dynamic filter has been updated since the initial check

Author: Done — added the `only when the filter has updated since the initial check` qualifier.

min/max can't possibly beat the threshold. The pruned row groups are
**never decoded, not even on the filter column**.

### Architecture · who drives the IO + decode loop

> <img src="/blog/images/sort-pushdown/arch_one_glance.png" alt="Three eras of who drives the parquet IO + decode loop" width="100%" class="img-fluid"/>

Reviewer: I am not sure that describing the evolution of the internal control loop implementation is important for the main point of the article -- it will take mental energy to understand the history / evolution; I think it would be easier to read if you just summarized the current state of the decoder loop (aka the "loop and decision point" section)

Author: Done — removed the three-eras evolution history; kept the current loop only.

The interesting backstory is that **DataFusion didn't actually own
this loop until recently**. Three eras:

* **Pre-[#20839]**: arrow-rs owned the I/O + decode loop as a black
  box; DataFusion only called `.next()` and served byte ranges. The
  row-group list was frozen at construction, so once the loop started,
  no mid-stream decisions were possible.
* **[#20839]**: the push-based parquet decoder moved the loop into
  DataFusion. The capability to insert a decision mid-loop now
  existed — but the loop went from `drain` straight to `drive`, with
  no decision point.
* **[#22450]**: adds the missing decision point. At every row-group
  boundary, the loop pauses to ask the runtime pruner whether the
  remaining row groups are still worth reading.

### The loop, and the decision point [#22450] adds

> <img src="/blog/images/sort-pushdown/transition_anatomy.png" alt="transition() loop: drain, decide, drive — Step 2 is the #22450 addition" width="100%" class="img-fluid" /><br/>

Reviewer: Some comments on the figure: 
<img width="978" height="507" alt="Image" src="https://github.com/user-attachments/assets/2545e195-3931-43b4-973b-579117f7c3c2" />

Author: Noted — figure annotation, will update the diagram separately.

Author: Done — regenerated `transition_anatomy.png` in 4e8a2d6. Widened the Step 2 / Step 3 code boxes so `into_builder().with_row_groups(...)` and `Data → set_active_reader` fit inside their boxes. Bottom captions removed (already covered in blog text).

*Figure: the decoder loop has three steps. Step 2 (DECIDE) is what
[#22450] adds — it only fires at row-group boundaries.*

The loop body reads: **drain** the current row group's batches until
it's exhausted; **decide** at the boundary whether any of the
remaining row groups can be dropped based on the live threshold; then
**drive** the decoder into the next row group and repeat. Inside a
row group, only drain and drive run — no decision point.

> <img src="/blog/images/sort-pushdown/pruner_loop.png" alt="RowGroupPruner: watch (cheap), rebuild (expensive, only if changed), prune (cheap)" width="100%" class="img-fluid" /><br/>

Reviewer: Sme comments: <img width="926" height="434" alt="Image" src="https://github.com/user-attachments/assets/c252125e-af2d-4a5b-95f9-abda60222e5d" />

Author: Noted — figure annotation, will update the diagram separately.

Author: Done — regenerated `pruner_loop.png` in 4e8a2d6. Widened the WATCH shape so `1 atomic load` fits without cutoff. Dropped the `Errors fall back to keep` line (too much detail per your comment).

*Figure: the pruner has a cheap "check if the filter changed" step, a
moderately expensive "rebuild the predicate if so" step, and a cheap
"apply the predicate to remaining row groups" step.*

The pruner is designed so the expensive work only fires when it can
possibly help: a cheap epoch check tells it whether the dynamic filter
has actually changed since last time, and only then does it rebuild
the pruning predicate. The predicate is then applied to remaining
row groups' min/max statistics — pure metadata comparison, no I/O.
Errors always fall back to "keep the row group" — a flaky pruner

> never drops live data.

Reviewer: You can probably tighten this up / make it more concise

```suggestion
The pruner is designed so the expensive work only fires when it can
possibly help: a cheap epoch check tells it whether the dynamic filter
has actually changed since last time, and only then does it rebuild
the pruning predicate. The predicate is then applied to remaining
row groups' min/max statistics, using only metadata, no I/O.
```

Author: Done — applied suggested phrasing.

> ### Cascading prune · how the heap eats row groups

Reviewer: I found this section confusing -- 

Some thoughts:
1. Call it "Example of pruning row groups using top k dynamic predicates"
2. Add an introductory sentence / parapgrah that explain what the section -- namely an example of how the techniques described in this blog work together to prune row groups. 

It might be too much for the blog, but I think the example would be clearer if there were a second figure that shows how Row Group Pruning works without the dynamic row group reordering -- that way the benefit of the approach in this blog might be easier to understand in comparsion to another fiure

Author: Done — renamed to `Example of pruning row groups using TopK dynamic predicates` + added intro.

every remaining row group with `max < 90` is pruned in one pass.*

The savings compound because the threshold moves in **steps**, not
smoothly. For `ORDER BY x DESC LIMIT 10` on a 10-row-group file
where reorder puts high-value row groups first:

1. RG 9 (values `[90..100)`) opens. One row group is enough to fill
   the heap of size 10 — the threshold jumps into RG 9's range (≥ 90).
2. At the next row-group boundary, the pruner sees that all of RG 8
   through RG 0 have `max < 90` and drops them in one pass.
3. Bytes for those nine row groups were **never fetched** — not
   projection columns, not the filter column. Full I/O + decompress +
   decode all skipped.

This is the unconditional value of [#22450]: when reorder lines up
disjoint per-RG ranges (the common case for time-series or
partition-key sorts), a single row group can cascade-eliminate every
remaining row group at the next boundary.

## Three-Layer Pruning · file + RG + row, stacked

<img src="/blog/images/sort-pushdown/pruning_stack.png" alt="Three-layer pruning: file-level, RG-level, row-level, all driven by the same TopK dynamic filter" width="100%" class="img-fluid"/>

> A common question at this point: "if [#22450] prunes whole row

Reviewer: I found this section somewhat confusing too as it seems to be illustrating the multi-level pruning strategy, but the words are descrobing the RowFilter level pruning

I would suggest moving this section earlier in the blog to introduce the concept of  multi-layer pruning (which is what the sort based optimizations really take advantage of)

Author: Done — reframed as an intro to the multi-layer pruning strategy.

Author: Follow-up on the reorder point — done in 9037e8a. Moved the Three-Layer Pruning section from after Runtime Row-Group Dynamic Pruning to right after How DataFusion Tracks Ordering, so the multi-layer framework is established before each layer is detailed. Promoted the topk_tpch benchmark to top-level ## since its parent section moved.

groups, does that replace the `RowFilter` row-level prune that the
`Inexact` path was already using?" **No** — the three layers stack,
and they're driven by the **same** `TopK` dynamic filter. (The
"Tier 1 / Tier 2" framing earlier maps to "Layer 0 / Layer A"
below — same partition, different lens. Layer B is what runs on
each row group after Layer A keeps it.)

> * **Layer 0 · file-level** (`file_pruner` + `EarlyStoppingStream`).

Reviewer: Is there a reason this list uses inconsistent labeling (`0`, `A`, and `B` rather than `1`, `2`, `3`, or `A`, `B`, `C`?

Author: Done — relabeled Layer 0/A/B → Layer 1/2/3 throughout.

  survive Layer A, the filter is still evaluated row-by-row to
  build a `RowSelection`. Rows that fail the predicate get their
  *projection* columns short-circuited via arrow-rs's
  `selects_any()`, but the *filter* column is necessarily read.
  This layer has the highest residual cost (the filter column),
  but also the finest granularity.

The same dynamic filter drives all three. A single insertion into
the `TopK` heap becomes a new threshold that Layer B applies
per-row immediately (in the currently-open row group), and Layer A
re-applies to remaining row groups at the next boundary. No layer
subsumes another — Layer A prunes on metadata alone (never touching
the filter column), while Layer B is finer-grained but has to read
the filter column to decide.

### Benchmark · `topk_tpch` (TPC-H SF1, `LIMIT 100`)

<img src="/blog/images/sort-pushdown/topk_tpch_bench.png" alt="topk_tpch benchmark results: 5 of 11 queries 3-4× faster, 0 regressions, total -44%" width="100%" class="img-fluid"/>

The [`topk_tpch`](https://github.com/apache/datafusion/blob/main/benchmarks/src/sort_tpch.rs) benchmark runs 11 TPC-H SF1 queries, all of the

> shape `ORDER BY ... LIMIT 100`, comparing `main` against the same

Reviewer: As the point of this blog is to explain these sort based optimizations, I recommend comparing:
1. Datafusion `54` without the sort optimizations (disable them)
2. DataFusion `54` with the sort optimizations

(This might be what it is doing, but the text says it is comparing the result of some particular PR)

That way the results are showing how effecive this approach is (rather than a particular PR)

Author: Done — reframed as DF54 sort-opt disabled vs enabled.

branch with [#22450] enabled. Headline numbers:

| Metric                              | Value                              |
| ----------------------------------- | ---------------------------------- |
| Total wall-clock (sum of 11 queries) | 248.8 ms → 139.1 ms (**−44%**)    |
| Queries with ≥2× speedup vs main    | **5 of 11** (Q2, Q4, Q8, Q9, Q10) |
| Queries with regression vs main     | **0**                              |
| Best single-query speedup           | **~4×**                            |

> The five queries with significant speedups all use `l_orderkey`

Reviewer: If you want to explain the results like this I suggest moving the details of the  tpch_sort benchmark to the intro

> The [`topk_tpch`](https://github.com/apache/datafusion/blob/main/benchmarks/src/sort_tpch.rs) benchmark runs 11 TPC-H SF1 queries, all of the
> shape `ORDER BY ... LIMIT 100`. The data is sorted by l_orderkey,  stored in parquet, with each row group XXX size , 1.5 distinct values,....

Author: Done — moved suite description into the intro.

as the **leading** sort key — lineitem's physical sort key, a
`BIGINT` with ~1.5M distinct values per SF1, so per-RG `min/max`
ranges are cleanly disjoint and `Layer A` can cascade-prune
aggressively. The non-winners (Q1, Q3, Q5, Q6, Q7, Q11) lead with
`l_linenumber` (cardinality 7), `l_comment`, or `l_shipmode` —
columns whose per-RG ranges overlap heavily because they're not the
physical sort order. (Q5–Q7 still *include* `l_orderkey`, but only
as a third-key tie-breaker — the leading key is what controls RG-level
disjointness.) A tighter threshold doesn't translate into clean
RG-level boundaries to prune at, so `Layer B` (row-level) still does
its share of the work.

> The takeaway isn't "5 out of 11", it's "**zero regressions and

Reviewer: This is confusing to me -- rather than say what the take away is NOT, how about just say what the result is -- several common timeseries workloads are accelerated 3-4x?

Author: Done — reframed positively: `3–4× on common time-series workloads with zero regressions`.

no-op when the data doesn't help, 3–4× when it does**". The sweet
spot — sort key aligned with the physical layout — is the common
case for time-series, partitioned tables, and ingestion-ordered
event logs.

> ## Future Directions

Reviewer: Given how long the blog already is, I suspect many readers will not study potential future work.  I recommend summarizing the future work in a sentence or two with links to the relevant arrow / datafusion tickets (and bonus points maybe to move these diagrams to those issues).

Author: Done — collapsed Future Directions to a summary sentence with issue links.

[#22450] prunes whole row groups at row-group boundaries. The
finer-grained extension prunes whole **pages** within a surviving
row group. The signal is the same dynamic filter, just re-applied
at page granularity — for any page whose `max(col)` is already
below the threshold, the filter column's bytes for that page can be
skipped along with the projection columns.

Today's page-level pruning runs once at file open using the static
query predicate. Future B extends [#22450]'s "refresh at RG
boundary" pattern to also rebuild the page-level filter with the
live threshold, so upcoming row groups get tighter page selections
mid-scan. Same arrow-rs API [#22450] already uses — no new
primitive needed. Tracked in [apache/datafusion#23216].

[apache/datafusion#23216]: https://github.com/apache/datafusion/issues/23216

Conceptually this is the same idea as [#22450] stepped down one
level: every level of the parquet hierarchy gets to chip off its
share of the residue from the level above.

## Acknowledgements

> Thank you to [@adriangb], [@alamb], [@xudong963], [@2010YOUY01], and

Reviewer: It might also be worth giving a shout out to Massive for helping / sponsoring your work if appropriate.

Author: Done — added a Massive shout-out in Acknowledgements.

* Morsel-style work scheduling: [apache/datafusion#21351](https://github.com/apache/datafusion/pull/21351)
* Runtime reorder for `TopK` convergence: [apache/datafusion#21956](https://github.com/apache/datafusion/pull/21956)
* **Runtime row-group dynamic pruning ([#22450])** — the centerpiece of this post.

In flight / open:

* Page-level reverse (arrow-rs): [apache/arrow-rs#9937](https://github.com/apache/arrow-rs/pull/9937), discussion in [apache/arrow-rs#9934](https://github.com/apache/arrow-rs/issues/9934)
* `peek_next_row_group` API for per-RG `fully_matched` RowFilter skip (arrow-rs): [apache/arrow-rs#10158](https://github.com/apache/arrow-rs/pull/10158)
* Page-level dynamic prune at RG boundary (Future B): [apache/datafusion#23216](https://github.com/apache/datafusion/issues/23216)
* Per-RG `fully_matched` RowFilter skip on top of [#22450] (blocked on arrow-rs#10158): [apache/datafusion#23067](https://github.com/apache/datafusion/issues/23067)
* Multi-column / function-wrapped stats reorder follow-ups: [apache/datafusion#22198](https://github.com/apache/datafusion/issues/22198)

> Concretely useful issues for new contributors:

Reviewer: I recommend also adding a "get involved" section: https://datafusion.apache.org/blog/2026/05/28/writing-agent-skills/#get-involved

Author: Done — added `Get Involved` section.

---
layout: post
title: Sort Pushdown in DataFusion: Skip Sorts, Skip Decode, Skip I/O
date: 2026-07-05
author: Qi Zhu
categories: [performance]
---

<!--
{% comment %}
Licensed to the Apache Software Foundation (ASF) under one or more
contributor license agreements.  See the NOTICE file distributed with
this work for additional information regarding copyright ownership.
The ASF licenses this file to you under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

[TOC]

*Qi Zhu, [Massive](https://www.massive.com/)*

> Many [Apache Parquet] datasets are already sorted on disk. Time-series

Reviewer: I suggest you start this blog with a 2-3 sentence abstract /  summary of its contents, so that a potential reader can quickly decide if they want to read it or not

Maybe something like

"DataFusion now automatically takes advantage of sortedess in the data, even when the data is only partially sorted and DataFusion does not know the sort order ahead of time. This blog explains why this is an important usecase and how Datafusion achieves this, through a combination of pushing down sort information, and adaptive runtime reordering and pruning"

Author: Done — added a 3-sentence abstract at the top.

> files are usually written in ingestion-time order. Event logs are sharded

Reviewer: Here are a few other motivating observations that might be worth working in
1. Data often has a partial sort order that is either not entirely sorted (e.g. ingestion order vs data order) or else the database doesn't know about the existing ordering
2. Modern data lakes based on Apache Icerberg and others often have to work with data as it is / was written and don't have the luxury of making a second copy, or resorting -- so being able to automatically exploit existing sortedness is important

Author: Added: time-series files, event logs, partition-keyed tables, and Iceberg / data-lake data-as-written are now called out.

and sorted by event id. Partitioned tables come with a natural ordering
implied by the partition key. The information about that ordering is
sitting right there in the file metadata.

[Apache Parquet]: https://parquet.apache.org/

[Apache DataFusion] has always been able to skip the sort in the

> *trivial* case: when the catalog declares an ordering (`WITH ORDER`

Reviewer: I am not sure I would call the code for the existing sort orderings "trivial" 😆  -- I would probably use the term "exact" -- so like "been able to skip the sort in the *exact* case"

Maybe you could also link to this blog from @akurmustafa that explains what the existing function is : https://datafusion.apache.org/blog/2025/03/11/ordering-analysis/

Reviewer: I also recommend  linking to the docs about `WITH ORDER` as it isn't SQL standard (it is a datafusion specific thing)

Author: Renamed to *exact* and linked to akurmustafa's ordering-analysis post.

Author: Added a link to the WITH ORDER docs.

or parquet `sorting_columns`) **and** the on-disk file listing
already matches that order, the `EnsureRequirements` rule sees that

> the scan's `output_ordering` already satisfies the request and

Reviewer: This is kind of a lot of detail -- I know what it means but many readers might not -- it might help to write this in non-datafusion technical terms. So instead of "when the catalog declares ..." something more like

"when the data on disk is perfectly sorted and table definition includes the ordering,  DataFusion  can avoid redundant sorts, as explained in the blog ...."

Author: Rewrote using plainer terms — no more `catalog declares`.

> **removes the redundant `SortExec`**. The hard cases were

Reviewer: And then in this next sentence, rather than calling it "hard" maybe call it "another" case and describe in terms of user visible behavior: "This blog is about taking advantage of sorting even when the files may not be perfectly sorted and DataFusion is not told about any ordering up front"

Author: Reframed as 'everything else — the messier real-world cases' with a bulleted list.

everything just slightly looser — files in the wrong on-disk order,
declared but overlapping ranges, no declaration at all,
`ORDER BY ... DESC` on ASC-sorted data — which all paid a full
external sort across the entire scan. CPU wasted. Memory wasted.
Streaming defeated.

[Apache DataFusion]: https://datafusion.apache.org/

> This post walks through the **sort pushdown** series of work — a

Reviewer: See overall comment about linear story telling vs problem/resolution structure

Author: Restructured the opening as **problem → techniques → results**, and cut the linear phase-by-phase framing.

multi-release effort spanning DataFusion **v52 through v55+** — that
closed that gap on three layers at once:

1. **`Exact` path** — delete the `SortExec` entirely when statistics
   prove the scan is already ordered. Headline: **2.1×–49×** on the
   `sort_pushdown` benchmark suite.
2. **`Inexact` path** — keep the `SortExec`, but read the

>    most-promising data first so `TopK`'s dynamic filter tightens fast

Reviewer: I suggest you link to the relevant blog post about topk / dynamic filters: https://datafusion.apache.org/blog/2025/09/10/dynamic-filters/

Author: Linked from the abstract and multiple in-body mentions of `TopK`'s dynamic filter.

   and stale row groups get pruned by statistics.
3. **Runtime row-group pruning** ([#22450]) — the latest piece, which
   re-checks the dynamic filter at every row-group boundary and
   physically removes pruned row groups from the open `RecordBatch`

>    stream. Headline: **5 of 11** `topk_tpch` queries run **3–4× faster**

Reviewer: I think you can make this more concise -- e.g. by not naming the benchmark suite (as you haven't yet introduced them yet anyways). Just saying "2.x-49x faster for certain benchmarks" is probablt enough

Author: Trimmed to `2×–49× faster on ASC-LIMIT queries` — dropped the suite name.

   with zero regressions; total benchmark time drops by **−44%**.

[#22450]: https://github.com/apache/datafusion/pull/22450

> [#20839]: https://github.com/apache/datafusion/pull/20839

Reviewer: Did we also merge the "reorder row groups to try and better take advantage of these optimizations" PR too?

Author: Yes — that's [#21956](https://github.com/apache/datafusion/pull/21956) (`Runtime reorder for TopK convergence`), listed under Landed PRs. Called it out more explicitly in the Inexact section.

> Together these compose into a **three-layer pruning stack** —

Reviewer: I think this paragrpah is redundant

Author: Removed.

file-level, row-group-level, row-level — all driven by the same
`TopK` dynamic filter. **None of the runtime parts of this post
would be possible without TopK's dynamic filter pushdown** — the
primitive that lets `TopK`'s threshold reach the parquet scanner
mid-execution. The earlier [Dynamic Filters][dyn-filters-blog]
post covers that primitive in detail and is recommended background
for what follows.

> The page-level reverse primitive we are adding upstream in

Reviewer: This note might be better left in a "future work" section at the end rather than in the intro

Author: Moved to Future Directions only. Intro no longer mentions it.

[arrow-rs] will push the `DESC` gains further still.

[arrow-rs]: https://github.com/apache/arrow-rs

> ## TL;DR

Reviewer: I don't think this is really a TLDR -- it is quite long and it also repeats information from the previous intro

<img width="1124" height="685" alt="Image" src="https://github.com/user-attachments/assets/3affadcd-d81d-4676-95a1-2220e34b64b3" />

I suggest removing this section and adding a 2-3 sentence summary at the very top

Author: Removed the TL;DR block entirely; the 3-sentence abstract at the top now covers it.

    `Inexact`, so `SortExec` stays in place, but `TopK`'s dynamic
    filter tightens fast and the rest is pruned. File-level early
    stop has been working for a while; **row-group-level early stop
    was the missing piece until [#22450]**.
  * **Runtime row-group dynamic pruning ([#22450])** — inside the
    decoder loop, at every row-group boundary, the live `TopK`
    threshold is converted into a fresh `PruningPredicate` and used
    to drop subsequent row groups before any bytes are fetched.
    A single row group can fill `TopK`'s heap, snap the threshold,
    and cascade-prune every remaining row group in one pass — no
    I/O, no decompress, no decode, not even the filter column.
* Benchmarks:
  * **`Exact` path on `sort_pushdown`**: `ORDER BY ... LIMIT` runs
    **27× and 49× faster**; full `ORDER BY` scans run **~2×** faster.
  * **[#22450] on `topk_tpch`** (TPC-H SF1, all `ORDER BY ... LIMIT 100`):
    5 of 11 queries get **3–4× faster**, **0 regressions**, total
    runtime drops **−44%** (248.8 ms → 139.1 ms).

## Why Sort Pushdown Matters

`SortExec` is one of the most expensive operators in a query plan.

> It is blocking by construction — no row can leave until every input

Reviewer: I think a more databasy term is "It is a pipeline blocker"

```rust
It is pipeline blocking operation  — no row can leave until every input
```

Author: That whole section was removed during the rewrite; the surviving mention uses `pipeline-blocking`.

row has been seen and compared — so it tends to dominate both latency
and peak memory. The cost gets paid even when:

* the file is already ordered by the sort key (very common for
  timestamp columns);
* the query only needs the top *N* rows (`ORDER BY ts LIMIT 100`), in
  which case full sort + truncate is wildly wasteful;

> * the next operator (`SortPreservingMergeExec`, `SortMergeJoinExec`,

Reviewer: I don't understand this point - those operators need sorted data so sorting isn't a wasted cost. If the data is already sorted then this would be covered by the first point

Author: Removed the confusing bullet during the intro rewrite.

  a window function) was going to consume ordered input anyway.

> The data DataFusion needs to avoid this work is **already in the file

Reviewer: It might also be useful to point out that it is common to use the statistics for filter based pruning, but 
that using them to infer sortedness / partial sortedness for sort based optimizations is not as common

Author: Added: 'Min/max statistics used for *predicate* pushdown are well-known... Using them to *reason about sort order* is less common.'

metadata**. Parquet writers can record per-column statistics (`min`,
`max`) at the row-group level. Files written by Spark, DuckDB,
arrow-rs, and others routinely include them. And explicit `WITH ORDER`

> clauses in DataFusion's SQL `CREATE EXTERNAL TABLE` give the optimizer

Reviewer: I thought this whole post was about taking advantge of sort order that Datafusion **didn't** already know about, and if so this mention of `WITH ORDER` is confusing

Author: Restructured so `WITH ORDER` only appears in the explicit 'what already worked' section, clearly separated from 'what was missing'.

a direct ordering hint. The job of sort pushdown is to **use that
information** — at plan time when possible, and at runtime when only
the live state of the query can tell us what's still worth reading.

## How DataFusion Tracks Ordering

<img src="/blog/images/sort-pushdown/plan-diff.svg" alt="EXPLAIN before / after: SortExec eliminated once ordering is Exact" width="100%" class="img-fluid"/>

> Each `FileScanConfig` carries an `output_ordering` — the ordering

Reviewer: If you are going to refer to DataFusion's internals, I recommend linking to the docs.rs or github url link. 

Also, as I mentioned above, I think this blog will have a larger audience if you minimize the DataFusion specific terminology as much as possible -- for example here you could say

> DataFusion tracks any pre-exisitng ordering using the `FileScanConfig::output_ordering` flag.
> The pre existing ordering can be either: 

and make sure to link to a definition of the `FileScanConfig`

Author: Linked inline to `docs.rs/FileScanConfig`. Also switched other internals to inline docs.rs / GitHub links throughout.

that the optimizer is willing to claim for the scan's output. There
are two flavours:

* **`Exact`** — the optimizer is *certain* the output is in this order.
  Optimizer rules treat an `Exact` ordering as a proof and **remove
  the surrounding `SortExec`** entirely. [`EnsureRequirements`] does
  this when the scan already declared `Exact` from the start; the
  [`PushdownSort`] rule covered in this post does the same after

>   upgrading from `Unsupported` via stats-based reorder.

Reviewer: This isa lot of deaail - I think this could be made a lot more concise by remvoing the implementation details. SOmething like 

**`Exact`** — the optimizer is *certain* the output is in this order and removes redundant `SortExec`s entirely. `

Author: Cut to a single sentence — implementation details dropped.

> * **`Inexact`** — the optimizer *believes* the output is probably

Reviewer: Suggest linking to the source / deifning wht these operators are

Author: Linked to docs.rs for `SortPreservingMergeExec`, `SortExec`, `FileScanConfig`, `BufferExec`, `FilePruner`, and `PushdownSort`.

  ordered, but cannot prove it. Downstream operators like
  `SortPreservingMergeExec` can still benefit from this hint, but the
  explicit `SortExec` stays for safety.

[`PushdownSort`]: https://github.com/apache/datafusion/blob/main/datafusion/physical-optimizer/src/pushdown_sort.rs

> A helper called `validated_output_ordering()` is the gatekeeper. It

Reviewer: This paragraph seems like an implementation detail and could be removed without changing the overall narrative

Author: Removed.

walks the list of files inside a partition, checks whether the
declared per-file ordering is consistent with the file order on disk,
and either confirms the ordering or **strips it entirely** if it
sees something ambiguous (e.g. file `b` comes before file `a` in the
file list but file `a`'s range comes first).

### `Exact` and `Inexact` at runtime

`Exact` and `Inexact` lead to different runtime behaviour, and
distinguishing them up front makes the rest of this post easier to
follow:

> * With **`Exact`**, the `SortExec` is removed and the LIMIT becomes

Reviewer: I found the mention of  "the `LIMIT`" here a little confusing as it wasn't previously referred to. 

Perhaps you could define an exemplar query like

> For example, let's consider a query such as the following to retrieve the 10 most recent trades:
> 
> ```sql
> SELECT timestamp, trade, amount ORDER BY timestamp DESC LIMIT 10
> ```
>
> Without any knowledge of the data sort order, DataFusion will scan all he data and
> use a Top-K operator. However, given sort knowledge:

Author: Added a concrete `SELECT ts, symbol, amount FROM trades ORDER BY ts DESC LIMIT 10` example with Exact / Inexact / no-info bullets.

  a **static fetch** on the source. The reader stops the moment the
  requested number of rows has been emitted — early termination
  at batch granularity, no dynamic state needed.
* With **`Inexact`**, the `SortExec` stays in place. The LIMIT
  materialises inside the sort as a `TopK` heap of size K. `TopK`
  exposes a [**dynamic filter**][dyn-filters-blog] — a runtime
  expression of the form *"only rows that could still beat the
  current K-th-best value are worth considering"* — and pushes it
  back to the parquet scanner. As more data is processed and the
  heap tightens, the filter's threshold tightens with it, and entire
  row groups can be skipped by checking the live threshold against
  the row group's min/max statistics. (See the earlier
  [dynamic filters][dyn-filters-blog] post for the full background
  on this mechanism.)

Both paths use the same underlying min/max statistics, but for
different purposes: `Exact` uses them at plan time to prove

> non-overlap and justify removing the sort; `Inexact` uses them at

Reviewer: It isn't clear to me from the text how Inexact uses the **same** min/max statistics... The Top-K / dynamc filter uses the same idea (track the maximum min) but i don't think it is the same statistics

Author: Reworded — no longer claims 'same statistics'. The two paths are now described separately (plan-time non-overlap proof vs. runtime pruning against a live heap threshold).

runtime to skip data that can no longer improve the heap.

[dyn-filters-blog]: https://datafusion.apache.org/blog/2025/09/10/dynamic-filters/

## The `PushdownSort` Rule

The **`PushdownSort`** physical optimizer rule asks each
`ExecutionPlan` two questions:

1. "Can you produce output in *this* ordering?"
2. "If yes, please rearrange yourself so that it actually does."

> The answer is one of `Exact`, `Inexact`, `Unsupported`. `Exact`

Reviewer: this seems somewhat redundant with the previous section --  I think if you combined them it would be clearer and more concise

Author: Merged the `PushdownSort` rule explanation into the tracking-ordering section as a single closing paragraph.

means the surrounding `SortExec` can be deleted entirely; `Inexact`
means the source will read the data in a near-sorted order so
`TopK` and other consumers benefit, but `SortExec` stays for
strict correctness; `Unsupported` means the source can't help and
`SortExec` does the full job. The rest of this post is what each
merged capability does on top of this protocol — first the `Exact`
path, then the `Inexact` path, then the runtime row-group prune
that `Inexact` was missing.

## The `Exact` Path · Sort Elimination via Statistics

<img src="/blog/images/sort-pushdown/phase1-file-reorder.svg" alt="Sort elimination: rearranging files within a partition by min/max statistics so the file list is in range order" width="100%" class="img-fluid"/>

A note on what's actually new here: the *trivial* sort-elimination
case (declared ordering + matching on-disk file list) has always
worked via the `EnsureRequirements` rule. What was missing was
sort elimination when the file list was in the **wrong order** on

> disk. `PushdownSort` landed across two phases — [Phase 1

Reviewer: I think the details of how the work landed could be omitted for brevity -- like maybe just say " What was missing was
sort elimination when the file list was in the **wrong order**, which was implemented in #19064 and #21182

Author: Now one sentence: `implemented across two PRs on PushdownSort: #19064 (scaffolding), #21182 (stats-based reorder)`.

(#19064)][#19064] introduced the rule scaffolding and reverse
iteration (the latter is used by the `Inexact` path covered
later); [Phase 2 (#21182)][#21182] added stats-based file reorder
so the rule can also produce `Exact` for the wrong-disk-order case.

> The motivating scenario for Phase 2:

Reviewer: "For example: "

Author: Done — `For example, consider three files a.parquet, b.parquet, c.parquet...`.

[#19064]: https://github.com/apache/datafusion/pull/19064
[#21182]: https://github.com/apache/datafusion/pull/21182

> * Three files: `a.parquet`, `b.parquet`, `c.parquet`.

Reviewer: this is a good example

Author: Thanks — kept the example.

* Each declares `WITH ORDER (ts ASC)`.
* Internally each file *is* sorted by `ts`.
* But they were written by different ingestion jobs and end up listed
  in the **wrong order** on disk (e.g. alphabetical by name, not by
  time).

> `validated_output_ordering()` looks at this, sees that the

Reviewer: I think this is a lot of implementation detail that will not mean much to anyone else -- just saying "previously DataFusion could not recognize the sort order, but now it can" is probably enough, with a link to the code for anyone who wants more details

Author: Cut the paragraph to essentials — three-step summary (sort file list / check adjacency / upgrade to Exact), with a link to `MinMaxStatistics` for readers who want more.

file-internal ordering disagrees with the file-list order, and
**strips the ordering entirely**. From the optimizer's point of view
the scan now has no declared ordering. Physical planning translates
the user's `ORDER BY` into a `SortExec`, and the earlier pipeline

> rule [`EnsureRequirements`] (which consolidates the old

Reviewer: BTW this doesnt' seem to be a link 

<img width="1000" height="403" alt="Image" src="https://github.com/user-attachments/assets/4afbee8f-98be-4fc1-9859-38430cddd339" />

Author: Fixed — all backtick-in-link patterns rewritten as inline `[X](url)` (10 total after a follow-up sweep). Verified in the rendered HTML: 0 literal `[X][ref]` remaining.

EnforceSorting + EnforceDistribution behaviour) can't remove it
because the scan can't prove the ordering. The data is sorted on
disk; the optimizer just can't tell.

[`EnsureRequirements`]: https://github.com/apache/datafusion/blob/main/datafusion/physical-optimizer/src/ensure_requirements/mod.rs

> Stats-based sort elimination fixes this in `PushdownSort`, which

Reviewer: This is also full of implementation details -- I suggest you focus on just explaining the conditions under which the transformation can be done (non overlapping min/max boundaries)

Author: Kept the phase2 non-overlap figure but stripped surrounding implementation detail; caption now just says `left case = safe to upgrade, right case = overlap, upgrade skipped, falls through to Inexact`.

   group**, not across groups; cross-group ordering is the job of
   `SortPreservingMergeExec` at runtime (more on this below).
3. **Upgrade `Unsupported` to `Exact`** when adjacency holds, the
   table has a declared `output_ordering` (from `WITH ORDER` or
   parquet `sorting_columns`), and the sort columns are null-free —
   the last condition preserves `NULLS LAST`/`NULLS FIRST` semantics
   across file boundaries. `PushdownSort` then removes the `SortExec`
   itself and the plan becomes streamable.

[`MinMaxStatistics`]: https://github.com/apache/datafusion/blob/main/datafusion/datasource/src/statistics.rs

One caveat that comes straight from `MinMaxStatistics`: the stats
sort only fires when every `ORDER BY` expression is a plain column.
`ORDER BY date_trunc('hour', ts)` silently skips the upgrade — there
is no per-file min/max for the function output to compare against.
Extending sort pushdown across monotonic function wrappers is one of
the open follow-ups.

(Runtime reorder covered later does handle some function-wrapped
sorts via monotonicity inference — but stats-based sort elimination
still needs a plain column.)

> <img src="/blog/images/sort-pushdown/phase2-stats-overlap.svg" alt="Detecting non-overlapping ranges via min/max statistics" width="100%" class="img-fluid"/>

Reviewer: The image is great -- I recommend adding a figure label to explain it to someone who is just skimming (not reading hte article linerly)

<img width="1000" height="403" alt="Image" src="https://github.com/user-attachments/assets/779c050a-4855-4b5f-a40a-aaec2c53e79f" />

Author: Done — added `*Figure: ...*` captions under every diagram.
